// typescript-logging/dist/commonjs/typescript-logging.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var LogGroupControl_1 = require("./control/LogGroupControl");
var CategoryServiceControl_1 = require("./control/CategoryServiceControl");
var ExtensionHelper_1 = require("./extension/ExtensionHelper");
exports.ExtensionHelper = ExtensionHelper_1.ExtensionHelper;
// Category related
var AbstractCategoryLogger_1 = require("./log/category/AbstractCategoryLogger");
exports.AbstractCategoryLogger = AbstractCategoryLogger_1.AbstractCategoryLogger;
var CategoryConsoleLoggerImpl_1 = require("./log/category/CategoryConsoleLoggerImpl");
exports.CategoryConsoleLoggerImpl = CategoryConsoleLoggerImpl_1.CategoryConsoleLoggerImpl;
var CategoryDelegateLoggerImpl_1 = require("./log/category/CategoryDelegateLoggerImpl");
exports.CategoryDelegateLoggerImpl = CategoryDelegateLoggerImpl_1.CategoryDelegateLoggerImpl;
var CategoryLogger_1 = require("./log/category/CategoryLogger");
exports.Category = CategoryLogger_1.Category;
var CategoryMessageBufferImpl_1 = require("./log/category/CategoryMessageBufferImpl");
exports.CategoryMessageBufferLoggerImpl = CategoryMessageBufferImpl_1.CategoryMessageBufferLoggerImpl;
var CategoryService_1 = require("./log/category/CategoryService");
exports.CategoryDefaultConfiguration = CategoryService_1.CategoryDefaultConfiguration;
exports.CategoryRuntimeSettings = CategoryService_1.CategoryRuntimeSettings;
exports.CategoryServiceFactory = CategoryService_1.CategoryServiceFactory;
var LoggerFactoryService_1 = require("./log/standard/LoggerFactoryService");
exports.LoggerFactoryOptions = LoggerFactoryService_1.LoggerFactoryOptions;
exports.LFService = LoggerFactoryService_1.LFService;
exports.LogGroupRule = LoggerFactoryService_1.LogGroupRule;
var AbstractLogger_1 = require("./log/standard/AbstractLogger");
exports.AbstractLogger = AbstractLogger_1.AbstractLogger;
var ConsoleLoggerImpl_1 = require("./log/standard/ConsoleLoggerImpl");
exports.ConsoleLoggerImpl = ConsoleLoggerImpl_1.ConsoleLoggerImpl;
var MessageBufferLoggerImpl_1 = require("./log/standard/MessageBufferLoggerImpl");
exports.MessageBufferLoggerImpl = MessageBufferLoggerImpl_1.MessageBufferLoggerImpl;
var LoggerOptions_1 = require("./log/LoggerOptions");
exports.CategoryLogFormat = LoggerOptions_1.CategoryLogFormat;
exports.DateFormat = LoggerOptions_1.DateFormat;
exports.DateFormatEnum = LoggerOptions_1.DateFormatEnum;
exports.LogFormat = LoggerOptions_1.LogFormat;
exports.LoggerType = LoggerOptions_1.LoggerType;
exports.LogLevel = LoggerOptions_1.LogLevel;
// Utilities
var DataStructures_1 = require("./utils/DataStructures");
exports.SimpleMap = DataStructures_1.SimpleMap;
exports.LinkedList = DataStructures_1.LinkedList;
__export(require("./utils/JSONHelper"));
var MessageUtils_1 = require("./utils/MessageUtils");
exports.MessageFormatUtils = MessageUtils_1.MessageFormatUtils;
/*
 Functions to export on TSL libarary var.
*/
// Export help function
function help() {
    /* tslint:disable:no-console */
    console.log("help()\n   ** Shows this help\n   \n getLogControl(): LoggerControl\n   ** Returns LoggerControl Object, use to dynamically change loglevels for log4j logging.\n   ** Call .help() on LoggerControl object for available options.\n   \n getCategoryControl(): CategoryServiceControl\n   ** Returns CategoryServiceControl Object, use to dynamically change loglevels for category logging.\n   ** Call .help() on CategoryServiceControl object for available options.   \n");
    /* tslint:enable:no-console */
}
exports.help = help;
// Export LogControl function (log4j)
function getLogControl() {
    return new LogGroupControl_1.LoggerControlImpl();
}
exports.getLogControl = getLogControl;
// Export CategoryControl function
function getCategoryControl() {
    return new CategoryServiceControl_1.CategoryServiceControlImpl();
}
exports.getCategoryControl = getCategoryControl;
